lib/async_hooks.js

```diff
@@ -197,6 +199,141 @@ class AsyncResource {
   }
 }
 
+function getResource(id) {
+  const weak = idResourcePairing.get(id);
+  if (!weak) {
+    return undefined;
+  }
+  return weak.get();
+}
+
+function getCurrentResource() {
+  const id = executionAsyncId();
+  return getResource(id);
+}
+
+const idResourcePairing = new Map();
+const contexts = [];
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // Save new resource.
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResource = getResource(triggerAsyncId);
+    if (triggerResource) {
+      for (let i = 0; i < contexts.length; ++i) {
+        contexts[i]._propagate(resource, triggerResource);
+      }
+    }
+  },
+  destroy(asyncId) {
+    const resource = getResource(asyncId);
+    if (resource) {
+      for (let i = 0; i < contexts.length; ++i) {
+        delete resource[contexts[i].kResourceStore];
+      }
+      idResourcePairing.delete(asyncId);
+    }
+  }
+});
+
+class AsyncContext {
+  constructor() {
+    this.kResourceStore = Symbol('kResourceStore');
+  }
+
+  /**
+   * Propagate the context from a parent resource to a child one.
+   */
+  _propagate(resource, triggerResource) {
+    const store = triggerResource[this.kResourceStore];
+    if (store) {
+      resource[this.kResourceStore] = store;
+    }
+  }
+
+  _enterSync() {
+    asyncContextHook.enable();
+    if (!contexts.includes(this)) {
+      contexts.push(this);
+    }
+    let resource = getCurrentResource();
+    if (!resource) {
+      // There is no resource, we need to create a fake one
+      resource = {};
+      idResourcePairing.set(executionAsyncId(), new WeakReference(resource));
+      process.nextTick(() => {
```

Comment:
This workaround may lead to context loss in `.runAndReturn()`. Let's consider the following snippet:
```js
const asyncContext = new AsyncContext();
asyncContext.runAndReturn(() => {
  asyncContext.getStore().set('foo', 'bar');
  // the following nextTick() is put into the queue after the _enterSync() one
  // and nothing prevents GC from collecting the fake resource
  process.nextTick(() => {
    // thus context may be lost here
    asyncContext.getStore().get('foo');
  });
});
```

Obviously, this issue may be resolved by migration to `executionAsyncResource()`.

Reply:
Please treat the logical chain above as an open question. I'd like to brainstorm if this workaround for fake resource may lead to context loss.

Reply:
That's a fair question. This is actually not possible:
The flow is the following:
* create fake resource
* create first nextTick
* create second nextTick <- at this point the context is attached to the second nextTick
* run first nextTick
* run nextTick

In my understanding, the only place where the loss is possible is if the first nextTick runs before we create the second one. Which does not happen.

I believe that playing around with the GC will show that this is not a problem here.

Reply:
```js
const {AsyncContext} = require('async_hooks')
const asyncContext = new AsyncContext();
asyncContext.runAndReturn(() => {
  asyncContext.getStore().set('foo', 'bar');
  gc();
  gc();
  gc();
  gc();
  gc();
  gc();
  process.nextTick(() => {
    // context may be lost here?
    console.log(asyncContext.getStore().get('foo'));
  });
});
```
gives me:
```
% ./node --expose-gc ttmp.js
bar
```
Do you see something I missed here?

Reply:
> This is actually not possible

Yes, you're right. This particular scenario can not happen. I wonder if this workaround may lead to unexpected behavior in other cases.

Going to think for a while on that topic. Feel free to resolve this thread.